dotCMS/src/main/java/com/dotmarketing/portlets/languagesmanager/business/LanguageFactoryImpl.java

```diff
@@ -552,117 +530,148 @@ private void saveLanguageKeys(String fileLangName, Map<String, String> keys, Set
-                    final String language    = resultSet.getString(4);
-                    final String country     = resultSet.getString(5);
-                    lang            = new Language(id, langCode, countryCode, language, country);
-                    CacheLocator.getLanguageCache().addLanguage(lang);
-                }
-            }
-        } catch (Exception e) {
 
-            Logger.error(LanguageFactoryImpl.class, "getLanguage failed:" + e, e);
-            throw new DotRuntimeException(e.toString());
-        } finally {
+		} catch (DotDataException e) {
 
-			closeQuietly(preparedStatement, resultSet);
-        }
+			Logger.error(LanguageFactoryImpl.class, "getLanguage failed:" + e, e);
+			throw new DotRuntimeException(e.toString());
+		}
 
-        return lang;
-    } // getFallbackLanguage.
+	} // getFallbackLanguage.
 
-    @Override
-    protected int deleteLanguageById(final Language language) {
+	@Override
+	protected int deleteLanguageById(final Language language) {
 
 		int rowsAffected = 0;
 		long id          = 0;
 
-        try {
+		try {
 
-            id = language.getId();
-            Logger.debug(this, "Deleting the language by id: " + id);
+			id = language.getId();
+			Logger.debug(this, "Deleting the language by id: " + id);
 			rowsAffected = this.dotConnect.executeUpdate(DELETE_FROM_LANGUAGE_WHERE_ID, id);
-        } catch (DotDataException e) {
-            Logger.error(LanguageFactoryImpl.class, "deleteLanguageById failed to delete the language with id: " + id, e);
-            throw new DotRuntimeException(e.toString(), e);
-        } finally {
+		} catch (DotDataException e) {
+			Logger.warn(LanguageFactoryImpl.class, "deleteLanguageById failed to delete the language with id: " + id);
+			throw new DotRuntimeException(e.toString(), e);
+		} finally {
 
 			CacheLocator.getLanguageCache().removeLanguage(language);
 		}
 
 		return rowsAffected;
-    } // deleteLanguageById.
+	} // deleteLanguageById.
+
+
+	private void dbUpsert(final Language language) throws DotDataException {
+
+		if (language.getId() == 0) {
+			language.setId(nextId());
+		}
+		Language tester = getLanguage(language.getId());
+		if (tester != null) {
+
+			new DotConnect().setSQL(UPDATE_LANGUAGE_BY_ID)
+					.addParam(language.getLanguageCode().toLowerCase())
+					.addParam(language.getCountryCode())
+					.addParam(language.getLanguage())
+					.addParam(language.getCountry())
+					.addParam(language.getId())
+					.loadResult();
+		} else {
+			DotConnect dc = new DotConnect().setSQL(INSERT_LANGUAGE_BY_ID)
+					.addParam(language.getId())
+					.addParam(language.getLanguageCode().toLowerCase())
+					.addParam(language.getCountryCode())
+					.addParam(language.getLanguage())
+					.addParam(language.getCountry());
+			dc.loadResult();
+		}
+		CacheLocator.getLanguageCache().removeLanguage(language);
+
+	}
+
+    private synchronized long nextId(){
+       return System.currentTimeMillis();
```

Comment:
I think it is same, even on cluster but eventually we can think on get the time millis from the db in order to do a lock to the db level, such as select getdate()

Reply:
yeah I totally agree.. in a clustered environment this would guarantee any unicity in the resulting id. And yes I've always thought that Dates should be gotten from the Database itself. But... 
1- This is what I was instructed to do. 
and 2. 
2 - You told me to use synchronized here.